agent/consul/gateways/controller_gateways.go

```diff
@@ -2,201 +2,571 @@ package gateways
+	// shared state. In our current out-of-repo code, this is handled via a global
+	// lock on our shared store, but this makes it so we don't have to deal with lock
+	// contention, and instead just work with a single control loop.
+	switch req.Kind {
+	case structs.APIGateway:
+		return reconcileEntry(r.fsm.State(), ctx, req, r.reconcileGateway, r.cleanupGateway)
+	case structs.BoundAPIGateway:
+		return reconcileEntry(r.fsm.State(), ctx, req, r.reconcileBoundGateway, r.cleanupBoundGateway)
+	case structs.HTTPRoute:
+		return reconcileEntry(r.fsm.State(), ctx, req, r.reconcileHTTPRoute, r.cleanupRoute)
+	case structs.TCPRoute:
+		return reconcileEntry(r.fsm.State(), ctx, req, r.reconcileTCPRoute, r.cleanupRoute)
+	case structs.InlineCertificate:
+		return r.enqueueCertificateReferencedGateways(r.fsm.State(), ctx, req)
+	default:
+		return nil
 	}
-	return controller.New(publisher, &reconciler).Subscribe(
-		&stream.SubscribeRequest{
-			Topic:   state.EventTopicAPIGateway,
-			Subject: stream.SubjectWildcard,
-		},
-	)
 }
 
-// Reconcile takes in a controller request and ensures this api gateways corresponding BoundAPIGateway exists and is
-// up to date
-func (r *apiGatewayReconciler) Reconcile(ctx context.Context, req controller.Request) error {
+func reconcileEntry[T structs.ControlledConfigEntry](store *state.Store, ctx context.Context, req controller.Request, reconciler func(ctx context.Context, req controller.Request, store *state.Store, entry T) error, cleaner func(ctx context.Context, req controller.Request, store *state.Store) error) error {
+	_, entry, err := store.ConfigEntry(nil, req.Kind, req.Name, req.Meta)
+	if err != nil {
+		return err
+	}
+	if entry == nil {
+		return cleaner(ctx, req, store)
+	}
+	return reconciler(ctx, req, store, entry.(T))
+}
 
-	r.logger.Debug("started reconciling gateway", "gateway", req.Name)
+func (r apiGatewayReconciler) enqueueCertificateReferencedGateways(store *state.Store, _ context.Context, req controller.Request) error {
+	logger := r.logger.With("inline-certificate", req.Name, "partition", req.Meta.PartitionOrDefault(), "namespace", req.Meta.NamespaceOrDefault())
+	logger.Debug("certificate changed, enqueueing dependent gateways")
+	defer logger.Debug("finished enqueuing gateways")
 
-	metaGateway, err := r.initGatewayMeta(req)
+	_, entries, err := store.ConfigEntriesByKind(nil, structs.APIGateway, acl.WildcardEnterpriseMeta())
 	if err != nil {
 		return err
-	} else if metaGateway == nil {
-		//delete meta gateway
-		r.logger.Info("cleaning up deleted gateway object", "request", req)
-		if err := r.store.Delete(&structs.BoundAPIGatewayConfigEntry{
-			Kind:           structs.BoundAPIGateway,
-			Name:           req.Name,
-			EnterpriseMeta: *req.Meta,
-		}); err != nil {
-			msg := "error cleaning up deleted gateway object"
-			r.logger.Error(msg, err)
-			return errors.Wrap(err, msg)
+	}
+	requests := []controller.Request{}
```

Comment:
#nit for readability

```suggestion
	}

	requests := []controller.Request{}
```